x-pack/filebeat/input/cometd/input_test.go

```diff
@@ -0,0 +1,376 @@
+
+		// close server
+		server.Close()
+	})
+
+	t.Run("Check gracefully shutdown input", func(t *testing.T) {
+		eventsCh := make(chan beat.Event)
+
+		outlet := &mockedOutleter{
+			onEventHandler: func(event beat.Event) bool {
+				eventsCh <- event
+				return true
+			},
+		}
+		connector := &mockedConnector{
+			outlet: outlet,
+		}
+		var inputContext finput.Context
+
+		var msg bay.MaybeMsg
+		msg.Msg.Data.Event.ReplayID = 1234
+		msg.Msg.Data.Payload = []byte(`{"CountryIso": "IN"}`)
+		msg.Msg.Channel = firstChannel
+
+		config := map[string]interface{}{
+			"channel_name":              firstChannel,
+			"auth.oauth2.client.id":     "client.id",
+			"auth.oauth2.client.secret": "client.secret",
+			"auth.oauth2.user":          "user",
+			"auth.oauth2.password":      "password",
+		}
+
+		r := http.HandlerFunc(oauth2Handler)
+		server := httptest.NewServer(r)
+		serverURL = server.URL
+		config["auth.oauth2.token_url"] = serverURL + "/token"
+
+		cfg := common.MustNewConfigFrom(config)
+
+		input, err := NewInput(cfg, connector, inputContext)
+		require.NoError(t, err)
+		require.NotNil(t, input)
+
+		// run input
+		input.Run()
+
+		go func() {
+			time.Sleep(time.Second) // let input.Stop() be executed.
+			input.Wait()
+		}()
+
+		start := time.Now()
+		for range []beat.Event{<-eventsCh} {
+			if time.Since(start) > time.Second {
+				require.Fail(t, "Channel was not closed after Wait()")
+			}
+		}
+	})
+
+	t.Run("Check input stop", func(t *testing.T) {
+		conf := defaultConfig()
+		logger := logp.NewLogger("test")
+		authParams := bay.AuthenticationParameters{}
+		inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+		workerCtx, workerCancel := context.WithCancel(inputCtx)
+		defer cancelInputCtx()
+
+		input := &cometdInput{
+			config:       conf,
+			log:          logger,
+			inputCtx:     inputCtx,
+			workerCtx:    workerCtx,
+			workerCancel: workerCancel,
+			authParams:   authParams,
+		}
+		input.msgCh = make(chan bay.MaybeMsg)
+
+		input.Stop()
+		select {
+		case <-workerCtx.Done():
+		default:
+			require.NoError(t, fmt.Errorf("input is not stopped."))
+		}
+	})
+
+	t.Run("Check input wait", func(t *testing.T) {
+		conf := defaultConfig()
+		logger := logp.NewLogger("test")
+		authParams := bay.AuthenticationParameters{}
+		inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+		workerCtx, workerCancel := context.WithCancel(inputCtx)
+		defer cancelInputCtx()
+
+		input := &cometdInput{
+			config:       conf,
+			log:          logger,
+			inputCtx:     inputCtx,
+			workerCtx:    workerCtx,
+			workerCancel: workerCancel,
+			authParams:   authParams,
+		}
+		input.msgCh = make(chan bay.MaybeMsg)
+
+		go func() {
+			time.Sleep(1000 * time.Millisecond)
+			input.Wait()
+		}()
+
+		time.Sleep(1000 * time.Millisecond) // let input.Stop() be executed.
+		select {
+		case <-workerCtx.Done():
+		default:
+			require.NoError(t, fmt.Errorf("input is not stopped."))
+		}
```

Comment:
```suggestion
		go func() {
			input.Wait()
		}()

		select {
		case <-workerCtx.Done():
		case <-time.After(time.Second): // let input.Stop() be executed.
			require.NoError(t, fmt.Errorf("input is not stopped."))
		}
```

The two delays being the same here might make this test flaky if run enough times. There's nothing really guarnateeing that the goroutine calling Wait() will unblock first.

I tested this change locally myself.